pkg/cloudprovider/providers/azure/azure_storage.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/arm/compute"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+)
+
+const (
+	maxLUN = 64 // max number of LUNs per VM
+)
+
+// AttachDisk attaches a vhd to vm
+// the vhd must exist, can be identified by diskName, diskURI, and lun.
+func (az *Cloud) AttachDisk(diskName, diskURI, vmName string, lun int32, cachingMode compute.CachingTypes) error {
+	vm, exists, err := az.getVirtualMachine(vmName)
+	if err != nil {
+		return err
+	} else if !exists {
+		return cloudprovider.InstanceNotFound
+	}
+	disks := *vm.Properties.StorageProfile.DataDisks
+	disks = append(disks,
+		compute.DataDisk{
+			Name: &diskName,
+			Vhd: &compute.VirtualHardDisk{
+				URI: &diskURI,
+			},
+			Lun:          &lun,
+			Caching:      cachingMode,
+			CreateOption: "attach",
+		})
+
+	newVM := compute.VirtualMachine{
+		Location: vm.Location,
+		Properties: &compute.VirtualMachineProperties{
+			StorageProfile: &compute.StorageProfile{
+				DataDisks: &disks,
+			},
+		},
+	}
+	_, err = az.VirtualMachinesClient.CreateOrUpdate(az.ResourceGroup, vmName, newVM, nil)
+	if err != nil {
+		glog.Errorf("azure attach failed, err: %v", err)
+		detail := err.Error()
+		if strings.Contains(detail, "Code=\"AcquireDiskLeaseFailed\"") {
+			// if lease cannot be acquired, immediately detach the disk and return the original error
+			glog.Infof("failed to acquire disk lease, try detach")
+			az.DetachDiskByName(diskName, diskURI, vmName)
+		}
+	} else {
+		glog.V(4).Infof("azure attach succeeded")
+	}
+	return err
+}
+
+// DetachDiskByName detaches a vhd from host
+// the vhd can be identified by diskName or diskURI
+func (az *Cloud) DetachDiskByName(diskName, diskURI, vmName string) error {
+	vm, exists, err := az.getVirtualMachine(vmName)
+	if err != nil || !exists {
+		// if host doesn't exist, no need to detach
+		glog.Errorf("cannot find node %s", vmName)
```

Comment:
Make this a warning and make it clear that detach is being skipped because of it.